### PR TITLE
[Site Design Revamp] Main View - Enlarge design thumbnails

### LINKED ISSUE #18448: [Site Design Revamp] Main View - Enlarge design thumbnails
## Description

Thumbnails under the [recommended category](https://github.com/wordpress-mobile/WordPress-iOS/issues/18434) should be enlarged to match the designs.

- On phones the designs should have a width of 200px
- On tablets the designs should have a width of 250px

| Before | After |
| ------- | ----- |
|  |  |

PRs should target the `feature/site-design-revamp` feature branch.

### PR DESCRIPTION
**Pull request by twstokes**
Fixes #18448

## Overview
- Allows `CategorySectionTableViewCell`s to have varying heights
- Enlarges thumbnails to match the project designs
  - See [SiteDesignCategoryThumbnailSize.swift](https://github.com/wordpress-mobile/WordPress-iOS/blob/task/site-design-enlarge-screenshots/WordPress/Classes/ViewRelated/Site%20Creation/Design%20Selection/SiteDesignCategoryThumbnailSize.swift) for values
- Adds support for mixing `CategorySection`s for various heights to prepare for #18434
- Sets iPhone Site Design thumbnails to `200x260`
- Sets iPad Site Design thumbnails to `250x325`

#### Site Design

| Before | After |
| - | - |
|  |  |

#### Page Layout (no changes expected)

| Before | After |
| - | - |
|  |  |

#### Site Design

| Before | After |
| - | - |
|  |  |

#### Page Layout (no changes expected)

| Before | After |
| - | - |
|  |  |

### This PR:
- Does not change the size of category labels.
- Does not address a large gap of space under the view's title. We can do this after adding the [recommended designs section](https://github.com/wordpress-mobile/WordPress-iOS/issues/18434).
- Does not address [this UITableView warning](https://github.com/wordpress-mobile/WordPress-iOS/issues/18514) (that predates the Site Design revamp project).

### Site Design picker

1. Start the Site Creation flow
2. Navigate to the Site Design picker

#### Expectations:
- "Ghost cells" are shown while the designs are loading and they fit the expected dimensions for iPhone and iPad (it can help to [slow down your network connection](https://nshipster.com/network-link-conditioner/))
- The thumbnails have dimensions of `200x260` on iPhone, and `250x325` on iPad
- Designs can be scrolled horizontally and vertically
- A row separator is shown between categories
- Rotating the view renders properly
- Tapping a design shows its preview
- Tapping "Choose" or "Create Site" (depending on Site Name experiment track) successfully creates a site with the chosen design
- Tapping "Skip" skips choosing a design
- (Running from Xcode only) No `UICollectionView` warnings are shown in the target output

### Page Design picker

1. Tap the FAB on the My Site view
2. Select "Site page"

#### Expectations:
- "Ghost cells" are shown while the designs are loading and they fit the expected dimensions for iPhone and iPad (it can help to [slow down your network connection](https://nshipster.com/network-link-conditioner/))
- The thumbnails have dimensions of `160x240` on both iPhone and iPad
- Designs can be scrolled horizontally and vertically
- A row separator is shown between categories
- Rotating the view renders properly
- Filtering categories works as expected
- (Running from Xcode only) No `UICollectionView` warnings are shown in the target output

## Regression Notes
1. Potential unintended areas of impact
  - Page Design picker (high risk)
    - Ghost views (shown when loading)
    - Filtering designs and scrolling
  - Site Creation flow

2. What I did to test those areas of impact (or what existing automated tests I relied on)
  - Manual testing across multiple devices.

3. What automated tests I added (or what prevented me from doing so)
  - No automated tests were added as this is very little logic and mostly UI code.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift:25]: Added so `CategorySections` can provide the size of their thumbnails (which all share the same size).

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift]: The cell height can now vary depending on the thumbnail size, so this now only provides the vertical padding that would be added to the cell.

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift]: This is clunky because of the way ghost thumbnails work - they aren't backed by any data source, so we can't find the height by querying the section (it will be nil).

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift]: There's probably no big win with this delegate method since we're not dealing with hundreds of rows, but I don't think it hurts to include it.

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift:35]: For recommended designs we should be able to do something similar by passing in a larger thumbnail size.

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift:64]: These had to be moved after `super.init`.

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.swift:37]: I first approached rows heights with a `UITableViewDelegate`, but the collection view's height would always be a little more than the thumbnail. Updating the constraint gave me deterministic heights, but I'm open to other ideas.

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift:35]: A use case for this is: I have a subclass and it has its own custom thumbnail size. When the content is loading I need to show ghost cells of the custom size, so I override this.

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.xib
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/CategorySectionTableViewCell.xib:55]: This may not have been necessary to change, but it seems to be a little more deterministic to always be 30.

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift
Comment (twstokes) [inline WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift]: @Gio2018 I realized this is a bug if for some reason the `ghostThumbnailSize` differs than what the actual content size is. 

Here's how it can happen, and why it currently only applies to future work when we add recommended designs:

1. Recommended designs will _probably_ be shown, so we set our ghost thumbnail size to the really big ones
2. We run into a scenario where we actually can't show recommended designs (there was no vertical chosen and no Blog category to fall back on)
3. When the `UITableView` goes to grab a new cell it properly sets the `cell.section`, which sets the right height for the content (a normal-sized cell)
4. We then immediately overwrite that height by setting the `ghostThumbnailSize` (our assumption from step 1) which is bigger, so we end up with too much vertical spacing around the thumbnail

One way to fix is is to just not set the ghost thumbnail size if we're not loading. I'll wait until your first review pass before making the update, unless I get your go-ahead before then. 👍

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift]: I think these two are not necessary, as in both cases the default value is `UITableView.automaticDimension` ([1](https://developer.apple.com/documentation/uikit/uitableview/1614852-rowheight), [2](https://developer.apple.com/documentation/uikit/uitableview/1614925-estimatedrowheight))

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignCategoryThumbnailSize.swift
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignCategoryThumbnailSize.swift:14]: I know it comes from the design specs, but 327 seems kind of a weird number here. Not a blocker for this PR at all, but @kyleaparker do you think we should round it up? Also it feels a bit odd that the recommended width on iPhone is larger than on iPad; I realize this is to obtain the effect of having one prominent recommended design on iPhone and two on iPad, but this might not be always the case depending on the actual screen size.

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift
Comment (Gio2018) [inline WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift]: Same observation as above applies here.

**Review by Gio2018 — APPROVED** [review]
Great work @twstokes ! I left a few non-blocking comments, but other than that :shipit: !

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift]: Good call @Gio2018 - thanks for catching that. I'll get these wiped. 👍

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignCategoryThumbnailSize.swift
Comment (kyleaparker) [reply WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignCategoryThumbnailSize.swift:14]: Yea the 327 comes from fitting exactly two designs within the space but I'm assuming it won't be possible to dynamically adjust the sizes? We could round to 300 x 400 to clean up the numbers and keep the ratio similar to the iPhone size 👍

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift]: This should be resolved with commit 447efa2bd0e74bb255e750a57c5f71b2858870e5.

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignContentCollectionViewController.swift]: This is resolved with commit 851270f66d95f25976b5656839e007402d2d4e16.

File: WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Gutenberg/Layout Picker/FilterableCategoriesViewController.swift]: This is resolved with commit 851270f66d95f25976b5656839e007402d2d4e16.

File: WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignCategoryThumbnailSize.swift
Comment (twstokes) [reply WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignCategoryThumbnailSize.swift:14]: 👋 @Gio2018 and @kyleaparker I'm going to go ahead and merge this PR into the feature branch with the current values. If we need to tweak them further we'll just make a new PR. Thanks!

Comment (Gio2018) [reply WordPress/Classes/ViewRelated/Site Creation/Design Selection/SiteDesignCategoryThumbnailSize.swift:14]: > but I'm assuming it won't be possible to dynamically adjust the sizes? 

we can adjust sizes. If we do so, we likely need to define how the sizes should behave when the screen width is larger than the height (landscape). In that case, I assume we probably would not want to fit them horizontally.
Numbers don't need to be rounded, it just seemed a bit unusual.

**Pull request merged**
